aten/src/ATen/native/mps/operations/FusedAdamWKernel.mm

```diff
@@ -65,4 +62,79 @@ void _fused_adamw_kernel_mps_(at::TensorList params,
                                 found_inf);
   }
 }
+
+void _fused_adamw_kernel_mps_(at::TensorList params,
+                              at::TensorList grads,
+                              at::TensorList exp_avgs,
+                              at::TensorList exp_avg_sqs,
+                              at::TensorList max_exp_avg_sqs,
+                              at::TensorList state_steps,
+                              const at::Tensor& lr,
+                              const double beta1,
+                              const double beta2,
+                              const double weight_decay,
+                              const double eps,
+                              const bool amsgrad,
+                              const bool maximize,
+                              const c10::optional<at::Tensor>& grad_scale,
+                              const c10::optional<at::Tensor>& found_inf) {
+  if (lr.is_cpu()) {
```

Comment:
I'm not familiar with foreach semantic, but don't you need to check that `lr` is scalar than? Because if `lr.is_cpu()` and it's not scalar, one shouldn't be taking `item` from it, should one? Also, wouldn't `item<double>()` throw an exception if `lr` dtype is say float32 or half?

Reply:
For the first point I think you're right, cc @janeyx99 as cpu and cuda are implemented the same way without the check.

For the second point, it does a type cast so no exception is thrown.

https://github.com/pytorch/pytorch/blob/fba5b7f7c8f050fb44bad6467ed19befbc9c9734/aten/src/ATen/templates/TensorMethods.cpp#L52-L56

Reply:
@malfet, I will address your first point in a follow up, as cpu and cuda have the same issue.

It actually fails at this line if lr's numel is not 1.

https://github.com/pytorch/pytorch/blob/fccbe593ed0c9b25271719054b046753345e5cea/torch/optim/adam.py#L46

`RuntimeError: Boolean value of Tensor with more than one value is ambiguous`